In [6]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [12]:
url='https://raw.githubusercontent.com/WalePhenomenon/climate_change/master/fuel_ferc1.csv'
fuel_df = pd.read_csv(url, error_bad_lines=False)
fuel_df.describe(include='all')

record_id  utility_id_ferc1   report_year  \
count                    29523      29523.000000  29523.000000   
unique                   29523               NaN           NaN   
top     f1_fuel_2005_12_57_3_9               NaN           NaN   
freq                         1               NaN           NaN   
mean                       NaN        118.601836   2005.806050   
std                        NaN         74.178353      7.025483   
min                        NaN          1.000000   1994.000000   
25%                        NaN         55.000000   2000.000000   
50%                        NaN        122.000000   2006.000000   
75%                        NaN        176.000000   2012.000000   
max                        NaN        514.000000   2018.000000   

       plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  \
count             29523               29523     29343     2.952300e+04   
unique             2315                   6         9              NaN   
top           big stone                 gas       mcf              NaN   
freq                156               11486     11354              NaN   
mean                NaN                 NaN       NaN     2.622119e+06   
std                 NaN                 NaN       NaN     9.118004e+06   
min                 NaN                 NaN       NaN     1.000000e+00   
25%                 NaN                 NaN       NaN     1.381700e+04   
50%                 NaN                 NaN       NaN     2.533220e+05   
75%                 NaN                 NaN       NaN     1.424034e+06   
max                 NaN                 NaN       NaN     5.558942e+08   

        fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
count          29523.000000               29523.000000   
unique                  NaN                        NaN   
top                     NaN                        NaN   
freq                    NaN                        NaN   
mean               8.492111                 208.649031   
std               10.600220                2854.490090   
min                0.000001                -276.080000   
25%                1.024000                   5.207000   
50%                5.762694                  26.000000   
75%               17.006000                  47.113000   
max              341.260000              139358.000000   

        fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
count                   2.952300e+04         29523.000000  
unique                           NaN                  NaN  
top                              NaN                  NaN  
freq                             NaN                  NaN  
mean                    9.175704e+02            19.304354  
std                     6.877593e+04          2091.540939  
min                    -8.749370e+02           -41.501000  
25%                     3.778500e+00             1.940000  
50%                     1.737100e+01             4.127000  
75%                     4.213700e+01             7.745000  
max                     7.964521e+06        359278.000000

In [13]:
fuel_df.head()

record_id  utility_id_ferc1  report_year  \
0   f1_fuel_1994_12_1_0_7                 1         1994   
1  f1_fuel_1994_12_1_0_10                 1         1994   
2   f1_fuel_1994_12_2_0_1                 2         1994   
3   f1_fuel_1994_12_2_0_7                 2         1994   
4  f1_fuel_1994_12_2_0_10                 2         1994   

       plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  \
0              rockport                coal       ton        5377489.0   
1  rockport total plant                coal       ton       10486945.0   
2                gorgas                coal       ton        2978683.0   
3                 barry                coal       ton        3739484.0   
4             chickasaw                 gas       mcf          40533.0   

   fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
0               16.590                      18.59   
1               16.592                      18.58   
2               24.130                      39.72   
3               23.950                      47.21   
4                1.000                       2.77   

   fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
0                         18.53                1.121  
1                         18.53                1.120  
2                         38.12                1.650  
3                         45.99                1.970  
4                          2.77                2.570

In [14]:
fuel_df.tail()

record_id  utility_id_ferc1  report_year  \
29518  f1_fuel_2018_12_12_0_13                12         2018   
29519   f1_fuel_2018_12_12_1_1                12         2018   
29520  f1_fuel_2018_12_12_1_10                12         2018   
29521  f1_fuel_2018_12_12_1_13                12         2018   
29522  f1_fuel_2018_12_12_1_14                12         2018   

           plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  \
29518    neil simpson ct #1                 gas       mcf          18799.0   
29519  cheyenne prairie 58%                 gas       mcf         806730.0   
29520     lange ct facility                 gas       mcf         104554.0   
29521       wygen 3 bhp 52%                coal       ton         315945.0   
29522       wygen 3 bhp 52%                 gas       mcf          17853.0   

       fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
29518                1.059                       4.78   
29519                1.050                       3.65   
29520                1.060                       4.77   
29521               16.108                       3.06   
29522                1.059                       0.00   

       fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
29518                          4.78                 9.03  
29519                          3.65                 6.95  
29520                          4.77                 8.99  
29521                         14.76                 1.11  
29522                          0.00                11.68

In [15]:
fuel_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29523 entries, 0 to 29522
Data columns (total 11 columns):
record_id                       29523 non-null object
utility_id_ferc1                29523 non-null int64
report_year                     29523 non-null int64
plant_name_ferc1                29523 non-null object
fuel_type_code_pudl             29523 non-null object
fuel_unit                       29343 non-null object
fuel_qty_burned                 29523 non-null float64
fuel_mmbtu_per_unit             29523 non-null float64
fuel_cost_per_unit_burned       29523 non-null float64
fuel_cost_per_unit_delivered    29523 non-null float64
fuel_cost_per_mmbtu             29523 non-null float64
dtypes: float64(5), int64(2), object(4)
memory usage: 2.5+ MB


## Q1
    

In [16]:
A = [1,2,3,4,5,6]
B = [13, 21, 34]


In [20]:
A.extend(B)

In [21]:
A

[1, 2, 3, 4, 5, 6, 13, 21, 34, 13, 21, 34, 13, 21, 34]

## Q2

In [22]:
np.identity(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

 ## Q3

In [24]:
fuel_df.head()


record_id  utility_id_ferc1  report_year  \
0   f1_fuel_1994_12_1_0_7                 1         1994   
1  f1_fuel_1994_12_1_0_10                 1         1994   
2   f1_fuel_1994_12_2_0_1                 2         1994   
3   f1_fuel_1994_12_2_0_7                 2         1994   
4  f1_fuel_1994_12_2_0_10                 2         1994   

       plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  \
0              rockport                coal       ton        5377489.0   
1  rockport total plant                coal       ton       10486945.0   
2                gorgas                coal       ton        2978683.0   
3                 barry                coal       ton        3739484.0   
4             chickasaw                 gas       mcf          40533.0   

   fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
0               16.590                      18.59   
1               16.592                      18.58   
2               24.130                      39.72   
3               23.950                      47.21   
4                1.000                       2.77   

   fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
0                         18.53                1.121  
1                         18.53                1.120  
2                         38.12                1.650  
3                         45.99                1.970  
4                          2.77                2.570

In [25]:
fuel_df.groupby('fuel_type_code_pudl')['fuel_cost_per_unit_burned'].mean()


fuel_type_code_pudl
coal         67.421830
gas          13.659397
nuclear    4955.157002
oil         168.877086
other        18.253856
waste        19.518122
Name: fuel_cost_per_unit_burned, dtype: float64

## The lowest average is Gas

## Q4

In [27]:
round(fuel_df.describe(),2)


utility_id_ferc1  report_year  fuel_qty_burned  fuel_mmbtu_per_unit  \
count          29523.00     29523.00     2.952300e+04             29523.00   
mean             118.60      2005.81     2.622119e+06                 8.49   
std               74.18         7.03     9.118004e+06                10.60   
min                1.00      1994.00     1.000000e+00                 0.00   
25%               55.00      2000.00     1.381700e+04                 1.02   
50%              122.00      2006.00     2.533220e+05                 5.76   
75%              176.00      2012.00     1.424034e+06                17.01   
max              514.00      2018.00     5.558942e+08               341.26   

       fuel_cost_per_unit_burned  fuel_cost_per_unit_delivered  \
count                   29523.00                      29523.00   
mean                      208.65                        917.57   
std                      2854.49                      68775.93   
min                      -276.08                       -874.94   
25%                         5.21                          3.78   
50%                        26.00                         17.37   
75%                        47.11                         42.14   
max                    139358.00                    7964521.00   

       fuel_cost_per_mmbtu  
count             29523.00  
mean                 19.30  
std                2091.54  
min                 -41.50  
25%                   1.94  
50%                   4.13  
75%                   7.74  
max              359278.00

## #fuel_mmbtu_per_unit the Std dev is 10.60 and 75th Percentile is 17.01.¶


## Q5

In [29]:
round(fuel_df.skew(),2)


utility_id_ferc1                  0.61
report_year                       0.01
fuel_qty_burned                  15.85
fuel_mmbtu_per_unit               4.14
fuel_cost_per_unit_burned        19.79
fuel_cost_per_unit_delivered    105.01
fuel_cost_per_mmbtu             171.68
dtype: float64

In [30]:
round(fuel_df.kurtosis(),2)


utility_id_ferc1                    1.09
report_year                        -1.15
fuel_qty_burned                   651.37
fuel_mmbtu_per_unit                55.60
fuel_cost_per_unit_burned         485.26
fuel_cost_per_unit_delivered    11765.05
fuel_cost_per_mmbtu             29489.13
dtype: float64

## Q6

In [32]:
feature = [col for col in fuel_df.columns if fuel_df[col].isnull().any()]
total = fuel_df['fuel_unit'].isnull().sum()
percent_missing = round(fuel_df['fuel_unit'].isnull().sum() * 100 / len(fuel_df),4)

In [33]:
print('Feature:',feature,', ''Total',total,', ''Percent',percent_missing)


Feature: ['fuel_unit'] , Total 180 , Percent 0.6097


## Q7

## The missing value for Feature falls under Categorical
### The missing value imputation technique to be used would be Mode Imputation



## Q8

In [34]:
fuel_df.corr()['fuel_cost_per_unit_burned'].sort_values(ascending=True)


utility_id_ferc1               -0.037863
fuel_qty_burned                -0.018535
fuel_mmbtu_per_unit            -0.010034
fuel_cost_per_mmbtu            -0.000437
fuel_cost_per_unit_delivered    0.011007
report_year                     0.013599
fuel_cost_per_unit_burned       1.000000
Name: fuel_cost_per_unit_burned, dtype: float64

### fuel_qty_burned and fuel_mmbtu_per_unit have second and third lowest correlation respectively with the Fuel Cost Per Unit Burned.


## Q9

In [35]:
fuel_df.head()


record_id  utility_id_ferc1  report_year  \
0   f1_fuel_1994_12_1_0_7                 1         1994   
1  f1_fuel_1994_12_1_0_10                 1         1994   
2   f1_fuel_1994_12_2_0_1                 2         1994   
3   f1_fuel_1994_12_2_0_7                 2         1994   
4  f1_fuel_1994_12_2_0_10                 2         1994   

       plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  \
0              rockport                coal       ton        5377489.0   
1  rockport total plant                coal       ton       10486945.0   
2                gorgas                coal       ton        2978683.0   
3                 barry                coal       ton        3739484.0   
4             chickasaw                 gas       mcf          40533.0   

   fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
0               16.590                      18.59   
1               16.592                      18.58   
2               24.130                      39.72   
3               23.950                      47.21   
4                1.000                       2.77   

   fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
0                         18.53                1.121  
1                         18.53                1.120  
2                         38.12                1.650  
3                         45.99                1.970  
4                          2.77                2.570

In [36]:
A= fuel_df.loc[(fuel_df['report_year']==1998) & (fuel_df['fuel_type_code_pudl']=='coal')]['fuel_cost_per_unit_burned'].sum()
A

11902.597

In [37]:
B= fuel_df.loc[(fuel_df['report_year']==1994) & (fuel_df['fuel_type_code_pudl']=='coal')]['fuel_cost_per_unit_burned'].sum()
B


14984.572000000002

In [38]:
print('The percentage change is:', round((A-B)*100/A))


The percentage change is: -26.0


## Q10

In [39]:
fuel_df.groupby('report_year')['fuel_cost_per_unit_delivered'].mean().sort_values(ascending=False)


report_year
1997    11140.197239
1996     9196.705948
2000      985.362877
2009      652.694163
2018      499.269966
2015      326.535511
1998      287.154420
2014      192.737183
2013      172.307591
2004      139.524275
2016      103.901761
2010       91.862105
1994       63.636060
2012       60.994502
2001       60.050396
2011       59.774667
2008       58.588197
2003       55.663493
2002       47.594361
2017       46.196861
2007       43.325023
2005       41.438184
2006       38.657484
1995       32.735269
1999       25.551627
Name: fuel_cost_per_unit_delivered, dtype: float64

### The year 1997 has the highest fuel_cost_per_unit_delivered